### ISSUE TITLE
"Could not solve these hints" error with USING INDEX

### ISSUE THREAD

**Issue by erobic** [neo4j/neo4j#12296]
I get a "Could not solve these hints" error with USING INDEX on MATCH command

- Neo4j version: 3.5.6
- Operating system: Ubuntu 16.04
- API/Driver: Cypher
- **Steps to reproduce**
The graph is defined as follows:

**Nodes: (each property is indexed)**
```
:Object(id, name)
:Attribute(name, concept_type)
```
**Edges:** Objects may point to Attribute using :HAS_ATTRIBUTE relation type.

**Query**
```
MATCH (o1:Object)-[:HAS_ATTRIBUTE]->(a1:Attribute)
WHERE a1.concept_type = 'color'
MATCH (o2:Object)-[:HAS_ATTRIBUTE]->(a2:Attribute)
USING INDEX o2:Object(name)
WHERE o1.name = o2.name
AND o1.id > o2.id
AND a1.name > a2.name
AND a1.concept_type = a2.concept_type
RETURN DISTINCT o1.name,a1.name,a2.name, a1.concept_type
LIMIT 1000
```
#### Expected Behavior:
Speed up query by using Object.name index

#### Actual Behavior:
I get the following error:

```
Could not solve these hints: `USING INDEX o2:Object(name)`

Plan Limit(SignedDecimalIntegerLiteral(1000), DoNotIncludeTies) {
  LHS -> Distinct(Map(o1.name -> Property(Variable(o1),PropertyKeyName(name)), a1.name -> Property(Variable(a1),PropertyKeyName(name)), a2.name -> Property(Variable(a2),PropertyKeyName(name)), a1.concept_type -> Property(Variable(a1),PropertyKeyName(concept_type)))) {
    LHS -> Selection(Ands(Set(Equals(Property(Variable(a1),PropertyKeyName(concept_type)),Property(Variable(a2),PropertyKeyName(concept_type))), Equals(Property(Variable(o1),PropertyKeyName(name)),Property(Variable(o2),PropertyKeyName(name))), AndedPropertyInequalities(Variable(a1),Property(Variable(a1),PropertyKeyName(name)),GreaterThan(Property(Variable(a1),PropertyKeyName(name)),Property(Variable(a2),PropertyKeyName(name))))))) {
      LHS -> Apply() {
        LHS -> Selection(Ands(Set(HasLabels(Variable(o2),List(LabelName(Object)))))) {
          LHS -> Expand(a2, INCOMING, List(RelTypeName(HAS_ATTRIBUTE)), o2,   UNNAMED101, ExpandAll) {
            LHS -> NodeByLabelScan(a2, LabelName(Attribute), Set()) {}
          }
        }
        RHS -> Selection(Ands(Set(In(Property(Variable(a1),PropertyKeyName(concept_type)),ListLiteral(List(Parameter(  AUTOSTRING0,String)))), HasLabels(Variable(a1),List(LabelName(Attribute)))))) {
          LHS -> Expand(o1, OUTGOING, List(RelTypeName(HAS_ATTRIBUTE)), a1,   UNNAMED18, ExpandAll) {
            LHS -> NodeUniqueIndexSeek(o1, LabelToken(Object,LabelId(3)), List(IndexedProperty(PropertyKeyToken(id,PropertyKeyId(8)),DoNotGetValue)), RangeQueryExpression(InequalitySeekRangeWrapper(RangeGreaterThan(ExclusiveBound(Property(Variable(o2),PropertyKeyName(id)))))), Set(o2, a2,   UNNAMED101), IndexOrderNone) {}
          }
        }
      }
    }
  }
}
```

Original forum discussions [here](https://community.neo4j.com/t/any-ideas-on-speeding-up-a-query-with-distinct-and-limit-already-in-place/10232/4).

**Comment by Hunterness** [comment]
Hi, thanks for reporting :) We will look into it.

/Team Cypher

**Comment by Hunterness** [comment]
Hi,

This was a bug where in some cases the hint was removed during planning and then failed when the chosen plan did not solve the hint. This has now been fixed and should be included in the next 3.5 release.

/Therese, team Cypher
